### PR TITLE
Stop throwing exception when writing 0 length buffers

### LINKED ISSUE #31299: ArgumentOutOfRangeException in ViewBufferTextWriter with latest servicing release
This line 
https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs#L113

throws when invoked with `index: 0, count: 0` which is the case when using the new html encoder `TextWriterExtensions`
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Encodings.Web/src/System/IO/TextWriterExtensions.cs#L33

when invoking this method with a string that starts with a chars that needs to be encoded:
https://github.com/dotnet/runtime/blob/1a35f1db908f96f2bddd9d23c780c515d14f16e0/src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs#L204

I have multiple reports of Orchard users with applications on Azure that stopped working when the release was pushed.
https://github.com/OrchardCMS/OrchardCore/issues/8887

### PR DESCRIPTION
**Pull request by TanayParikh**
Fixes: https://github.com/dotnet/aspnetcore/issues/31299

### CODE REVIEW

File: src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs [hunk 1]
```diff
@@ -109,6 +109,11 @@ public override void Write(char[] buffer, int index, int count)
             {
                 throw new ArgumentNullException(nameof(buffer));
             }
+            
+            if (index == 0 && count == 0)
```
Comment (TanayParikh) [inline src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs:113]: Not sure of the exact approach / level of complexity we'd like to add here. We could do something a bit fancier like:

```suggestion
            if (index >= 0 && index < buffer.Length && count == 0)
```

Comment (pranavkm) [reply src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs:113]: We could simply copy the checks from here: https://github.com/dotnet/runtime/blob/7938f9d9fadc3952d792329be602668bef124f23/src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs#L159

Comment (TanayParikh) [reply src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs]: Thanks, updated & added a test.

**Review by sebastienros — APPROVED** [review]

File: src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs [hunk 2]
```diff
@@ -110,16 +110,21 @@ public override void Write(char[] buffer, int index, int count)
...
                 throw new ArgumentOutOfRangeException(nameof(index));
             }
 
-            if (count < 0 || (buffer.Length - index < count))
+            if (count < 0)
             {
                 throw new ArgumentOutOfRangeException(nameof(count));
             }
 
+            if (buffer.Length - index < count)
```
Comment (pranavkm) [inline src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBufferTextWriter.cs:123]: Hmm, isn’t this basically the same check as we had before?

**Review by pranavkm — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
See also my comment at https://github.com/dotnet/aspnetcore/issues/31299#issuecomment-808766264. You may be leaving some performance on the floor given the current method implementations.

**Comment by TanayParikh** [comment]
> See also my comment at [#31299 (comment)](https://github.com/dotnet/aspnetcore/issues/31299#issuecomment-808766264). You may be leaving some performance on the floor given the current method implementations.

Thanks @GrabYourPitchforks, I created https://github.com/dotnet/aspnetcore/issues/32284 to track that separately.

**Pull request merged**